base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl

```diff
@@ -0,0 +1,1604 @@
+  * `x.AliasInfo::Unindexable` records all the possible values that can be aliased to fields/elements of `x` without precise index information
+- `x.Liveness::BitSet`: records SSA statement numbers where `x` should be live, e.g.
+  to be used as call argument, to be returned to the caller, or preserved for `:foreigncall`.
+  `0 ∈ x.Liveness` has the special meaning that it's used as a call argument of the analyzed.
+- `x.ArgEscape::Int` (not implemented yet): indicates it will escape to the caller through
+  `setfield!` on argument(s)
+  * `-1` : no escape
+  * `0` : unknown or multiple
+  * `n` : through argument N
+
+There are utility constructors to create common `EscapeLattice`s, e.g.,
+- `NoEscape()`: the bottom(-like) element of this lattice, meaning it won't escape to anywhere
+- `AllEscape()`: the topmost element of this lattice, meaning it will escape to everywhere
+
+`analyze_escapes` will transition these elements from the bottom to the top,
+in the same direction as Julia's native type inference routine.
+An abstract state will be initialized with the bottom(-like) elements:
+- the call arguments are initialized as `ArgEscape()`, because they're visible from a caller immediately
+- the other states are initialized as `NotAnalyzed()`, which is a special lattice element that
+  is slightly lower than `NoEscape`, but at the same time doesn't represent any meaning
+  other than it's not analyzed yet (thus it's not formally part of the lattice).
+"""
+struct EscapeLattice
+    Analyzed::Bool
+    ReturnEscape::Bool
+    ThrownEscape::LivenessSet
+    AliasInfo #::Union{Indexable,Unindexable,Bool}
+    Liveness::LivenessSet
+    # TODO: ArgEscape::Int
+
+    function EscapeLattice(
+        Analyzed::Bool,
+        ReturnEscape::Bool,
+        ThrownEscape::LivenessSet,
+        AliasInfo#=::Union{Indexable,Unindexable,Bool}=#,
+        Liveness::LivenessSet,
+        )
+        @nospecialize AliasInfo
+        return new(
+            Analyzed,
+            ReturnEscape,
+            ThrownEscape,
+            AliasInfo,
+            Liveness,
+            )
+    end
+    function EscapeLattice(
+        x::EscapeLattice,
+        # non-concrete fields should be passed as default arguments
+        # in order to avoid allocating non-concrete `NamedTuple`s
+        AliasInfo#=::Union{Indexable,Unindexable,Bool}=# = x.AliasInfo;
+        Analyzed::Bool = x.Analyzed,
+        ReturnEscape::Bool = x.ReturnEscape,
+        ThrownEscape::LivenessSet = x.ThrownEscape,
+        Liveness::LivenessSet = x.Liveness,
+        )
+        @nospecialize AliasInfo
+        return new(
+            Analyzed,
+            ReturnEscape,
+            ThrownEscape,
+            AliasInfo,
+            Liveness,
+            )
+    end
+end
+
+# precomputed default values in order to eliminate computations at each callsite
+const BOT_THROWN_ESCAPE = LivenessSet()
+const TOP_THROWN_ESCAPE = LivenessSet(1:100_000)
```

Comment:
`EscapeAnalysis` has many special cases in `EscapeInfo` lattice operations to avoid operations on those sentinels, e.g. join of `ThrownEscape` is implemented as like https://github.com/JuliaLang/julia/blob/b7171d4aaa357f717a4080a0699b31d6d1bb7d08/base/compiler/ssair/EscapeAnalysis/EscapeAnalysis.jl#L356-L365
So I think it actually never be involved with actual set computations.
Having said that, I think changing these sentinel to `Bool` would make those special cases explicit and makes sure we never do unnecessary computations.

Reply:
Actually, the current implementation seems to be more optimal.
With the current implementation, we get:
```
❯ jld -e 'using EscapeAnalysis; code_escapes(sin, (Int,)); @time code_escapes(Base.init_stdio, (Ptr{Cvoid},)); @time code_escapes(println, (QuoteNode,))'
  3.989914 seconds (24.00 M allocations: 1.261 GiB, 20.80% gc time, 90.39% compilation time)
  0.337571 seconds (2.36 M allocations: 128.885 MiB, 14.96% gc time, 99.98% compilation time)
```

But we get the following result (with a slight increase of allocations) if we change to `EscapeInfo.ThrownEscape::Union{Bool,LivenessSet}`
```
❯ jld -e 'using EscapeAnalysis; code_escapes(sin, (Int,)); @time code_escapes(Base.init_stdio, (Ptr{Cvoid},)); @time code_escapes(println, (QuoteNode,))'
  4.290117 seconds (24.09 M allocations: 1.266 GiB, 19.94% gc time, 90.36% compilation time)
  0.362981 seconds (2.37 M allocations: 129.322 MiB, 20.66% gc time, 99.98% compilation time)
```

The diff I tried is:
```diff
diff --git a/src/EscapeAnalysis.jl b/src/EscapeAnalysis.jl
index 91f0dd8..5870af7 100644
--- a/src/EscapeAnalysis.jl
+++ b/src/EscapeAnalysis.jl
@@ -120,7 +120,7 @@ An abstract state will be initialized with the bottom(-like) elements:
 struct EscapeInfo
     Analyzed::Bool
     ReturnEscape::Bool
-    ThrownEscape::LivenessSet
+    ThrownEscape::Union{Bool,LivenessSet}
     AliasInfo #::Union{IndexableFields,IndexableElements,Unindexable,Bool}
     Liveness::LivenessSet
     # TODO: ArgEscape::Int
@@ -128,11 +128,11 @@ struct EscapeInfo
     function EscapeInfo(
         Analyzed::Bool,
         ReturnEscape::Bool,
-        ThrownEscape::LivenessSet,
+        ThrownEscape::Union{Bool,LivenessSet},
         AliasInfo#=::Union{IndexableFields,IndexableElements,Unindexable,Bool}=#,
         Liveness::LivenessSet,
         )
-        @nospecialize AliasInfo
+        @nospecialize AliasInfo ThrownEscape
         return new(
             Analyzed,
             ReturnEscape,
@@ -148,7 +148,7 @@ struct EscapeInfo
         AliasInfo#=::Union{IndexableFields,IndexableElements,Unindexable,Bool}=# = x.AliasInfo;
         Analyzed::Bool = x.Analyzed,
         ReturnEscape::Bool = x.ReturnEscape,
-        ThrownEscape::LivenessSet = x.ThrownEscape,
+        @nospecialize(ThrownEscape::Union{Bool,LivenessSet} = x.ThrownEscape),
         Liveness::LivenessSet = x.Liveness,
         )
         @nospecialize AliasInfo
@@ -163,9 +163,6 @@ struct EscapeInfo
 end
 
 # precomputed default values in order to eliminate computations at each callsite
-const BOT_THROWN_ESCAPE = LivenessSet()
-const TOP_THROWN_ESCAPE = LivenessSet(1:100_000)
-
 const BOT_ALIAS_INFO = false
 const TOP_ALIAS_INFO = true
 
@@ -174,28 +171,28 @@ const TOP_LIVENESS = LivenessSet(0:100_000)
 const ARG_LIVENESS = LivenessSet(0)
 
 # the constructors
-NotAnalyzed() = EscapeInfo(false, false, BOT_THROWN_ESCAPE, BOT_ALIAS_INFO, BOT_LIVENESS) # not formally part of the lattice
-NoEscape() = EscapeInfo(true, false, BOT_THROWN_ESCAPE, BOT_ALIAS_INFO, BOT_LIVENESS)
-ArgEscape() = EscapeInfo(true, false, BOT_THROWN_ESCAPE, TOP_ALIAS_INFO, ARG_LIVENESS) # TODO allow interprocedural alias analysis?
-ReturnEscape(pc::Int) = EscapeInfo(true, true, BOT_THROWN_ESCAPE, BOT_ALIAS_INFO, LivenessSet(pc))
-AllReturnEscape() = EscapeInfo(true, true, BOT_THROWN_ESCAPE, BOT_ALIAS_INFO, TOP_LIVENESS)
+NotAnalyzed() = EscapeInfo(false, false, false, BOT_ALIAS_INFO, BOT_LIVENESS) # not formally part of the lattice
+NoEscape() = EscapeInfo(true, false, false, BOT_ALIAS_INFO, BOT_LIVENESS)
+ArgEscape() = EscapeInfo(true, false, false, TOP_ALIAS_INFO, ARG_LIVENESS) # TODO allow interprocedural alias analysis?
+ReturnEscape(pc::Int) = EscapeInfo(true, true, false, BOT_ALIAS_INFO, LivenessSet(pc))
+AllReturnEscape() = EscapeInfo(true, true, false, BOT_ALIAS_INFO, TOP_LIVENESS)
 ThrownEscape(pc::Int) = EscapeInfo(true, false, LivenessSet(pc), BOT_ALIAS_INFO, BOT_LIVENESS)
-AllEscape() = EscapeInfo(true, true, TOP_THROWN_ESCAPE, TOP_ALIAS_INFO, TOP_LIVENESS)
+AllEscape() = EscapeInfo(true, true, true, TOP_ALIAS_INFO, TOP_LIVENESS)
 
 const ⊥, ⊤ = NotAnalyzed(), AllEscape()
 
 # Convenience names for some ⊑ₑ queries
-has_no_escape(x::EscapeInfo) = !x.ReturnEscape && isempty(x.ThrownEscape) && 0 ∉ x.Liveness
+has_no_escape(x::EscapeInfo) = !x.ReturnEscape && x.ThrownEscape === false && 0 ∉ x.Liveness
 has_arg_escape(x::EscapeInfo) = 0 in x.Liveness
 has_return_escape(x::EscapeInfo) = x.ReturnEscape
 has_return_escape(x::EscapeInfo, pc::Int) = x.ReturnEscape && pc in x.Liveness
-has_thrown_escape(x::EscapeInfo) = !isempty(x.ThrownEscape)
-has_thrown_escape(x::EscapeInfo, pc::Int) = pc in x.ThrownEscape
+has_thrown_escape(x::EscapeInfo) = x.ThrownEscape !== false
+has_thrown_escape((; ThrownEscape)::EscapeInfo, pc::Int) = ThrownEscape === true || (isa(ThrownEscape, LivenessSet) && pc in ThrownEscape)
 has_all_escape(x::EscapeInfo) = ⊤ ⊑ₑ x
 
 # utility lattice constructors
 ignore_argescape(x::EscapeInfo) = EscapeInfo(x; Liveness=delete!(copy(x.Liveness), 0))
-ignore_thrownescapes(x::EscapeInfo) = EscapeInfo(x; ThrownEscape=BOT_THROWN_ESCAPE)
+ignore_thrownescapes(x::EscapeInfo) = EscapeInfo(x; ThrownEscape=false)
 ignore_aliasinfo(x::EscapeInfo) = EscapeInfo(x, BOT_ALIAS_INFO)
 ignore_liveness(x::EscapeInfo) = EscapeInfo(x; Liveness=BOT_LIVENESS)
 
@@ -207,10 +204,10 @@ x::EscapeInfo == y::EscapeInfo = begin
     x.Analyzed === y.Analyzed || return false
     x.ReturnEscape === y.ReturnEscape || return false
     xt, yt = x.ThrownEscape, y.ThrownEscape
-    if xt === TOP_THROWN_ESCAPE
-        yt === TOP_THROWN_ESCAPE || return false
-    elseif yt === TOP_THROWN_ESCAPE
-        return false # x.ThrownEscape === TOP_THROWN_ESCAPE
+    if isa(xt, Bool)
+        xt === yt || return false
+    elseif isa(yt, Bool)
+        return false # !isa(xt, Bool)
     else
         xt == yt || return false
     end
@@ -259,9 +256,13 @@ x::EscapeInfo ⊑ₑ y::EscapeInfo = begin
     x.Analyzed ≤ y.Analyzed || return false
     x.ReturnEscape ≤ y.ReturnEscape || return false
     xt, yt = x.ThrownEscape, y.ThrownEscape
-    if xt === TOP_THROWN_ESCAPE
-        yt !== TOP_THROWN_ESCAPE && return false
-    elseif yt !== TOP_THROWN_ESCAPE
+    if isa(xt, Bool)
+        if xt === true
+            yt !== true && return false
+        end
+    elseif isa(yt, Bool)
+        yt === false && return false # xt !== false
+    else
         xt ⊆ yt || return false
     end
     xa, ya = x.AliasInfo, y.AliasInfo
@@ -354,12 +355,18 @@ x::EscapeInfo ⊔ₑ y::EscapeInfo = begin
         return x
     end
     xt, yt = x.ThrownEscape, y.ThrownEscape
-    if xt === TOP_THROWN_ESCAPE || yt === TOP_THROWN_ESCAPE
-        ThrownEscape = TOP_THROWN_ESCAPE
-    elseif xt === BOT_THROWN_ESCAPE
-        ThrownEscape = yt
-    elseif yt === BOT_THROWN_ESCAPE
-        ThrownEscape = xt
+    if isa(xt, Bool)
+        if xt === true
+            ThrownEscape = true
+        else
+            ThrownEscape = yt
+        end
+    elseif isa(yt, Bool)
+        if yt === true
+            ThrownEscape = true
+        else
+            ThrownEscape = xt
+        end
     else
         ThrownEscape = xt ∪ yt
     end
@@ -1021,7 +1028,13 @@ function escape_exception!(astate::AnalysisState, tryregions::Vector{UnitRange{I
     for i in 1:length(escapes)
         x = escapes[i]
         xt = x.ThrownEscape
-        xt === TOP_THROWN_ESCAPE && @goto propagate_exception_escape # fast pass
+        if isa(xt, Bool)
+            if xt === true
+                @goto propagate_exception_escape
+            else
+                continue
+            end
+        end
         for pc in x.ThrownEscape
             for region in tryregions
                 pc in region && @goto propagate_exception_escape # early break because of AllEscape
@@ -1079,7 +1092,7 @@ the return value and `pc` is the SSA statement number of the return value.
 function from_interprocedural(arginfo::ArgEscapeInfo, retinfo::EscapeInfo, pc::Int)
     arginfo.AllEscape && return ⊤
 
-    ThrownEscape = arginfo.ThrownEscape ? LivenessSet(pc) : BOT_THROWN_ESCAPE
+    ThrownEscape = arginfo.ThrownEscape ? LivenessSet(pc) : false
 
     return EscapeInfo(
         #=Analyzed=#true, #=ReturnEscape=#false, ThrownEscape,
```

Reply:
We could also have `LivenessSet(1)` mean top, and otherwise `LivenessSet(2:nargs+1)` and `LivenessSet(2+nargs:nssavalues)` be the rest

Reply:
Since `LivenessSet` accepts negative integers, I think it might be simpler if we have `TOP_THROWN_ESCAPE = LivenessSet(-1)`.

Reply:
79333fb